### PR TITLE
Tim sort in C++

### PR DESCRIPTION
**Pull request by Harikrishnan6336**
Fixex  #415  in C++

### CODE REVIEW

**Comment by Harikrishnan6336** [comment]
Kindly review this PR @tarun26091999  @rudrakshi99

**Comment by Harikrishnan6336** [comment]
Kindly review this PR @raksha009

File: Code/C++/tim_sort.cpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
+/*
+    Tim Sort
+
```
Comment (raksha009) [inline Code/C++/tim_sort.cpp:3]: Remove extra line

File: Code/C++/tim_sort.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
+/*
+    Tim Sort
+    
+    Tim Sort is a stable sorting algorithm that combines the capabilities of  Merge sort     and Insertion sort.
+    It first splits the array into  small segments and sort them using Insertion Sort,       and then merges the small segments using merge operation.
+    It is used in Java’s Arrays.sort() as well as Python’s sorted() and sort()
+    
+    Time Complexity- O(n*log(n))
```
Comment (raksha009) [inline Code/C++/tim_sort.cpp:8]: Write Time and Space complexity at the end.

**Review by raksha009 — CHANGES_REQUESTED** [review]

**Comment by Harikrishnan6336** [comment]
@raksha009 I have made the suggested changes, Kindly review my PR.

File: Code/C++/tim_sort.cpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        } 
+        arr[j+1] = temp; 
+    } 
+} 
+
+
+// Merge operation to merge two sorted segments
+void merge(int arr[], int beg, int mid, int end)
+{
+
```
Comment (raksha009) [inline Code/C++/tim_sort.cpp:35]: Remove this extra line.

File: Code/C++/tim_sort.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    }
+}
+
+// Sort the array using Tim Sort
+void tim_sort(int arr[], int n) 
+{ 
+    // Sort individual subarrays of size SIZE 
+    for (int i = 0; i < n; i+=SIZE) 
+        insertion_sort(arr, i, min((i+15), 
+                                    (n-1)));
```
Comment (raksha009) [inline Code/C++/tim_sort.cpp:80]: plz combine line 79 and 80.

File: Code/C++/tim_sort.cpp [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        insertion_sort(arr, i, min((i+15), 
+                                    (n-1))); 
+    
+    // Merge all the small segments formed
+    for (int size = SIZE; size < n; 
+                            size = 2*size) 
+    { 
+    // From the start of the array, we are going to merge two segments each of size SIZE.
+        for (int left = 0; left < n; 
+                            left += 2*size)
```
Comment (raksha009) [inline Code/C++/tim_sort.cpp:88]: plz combine line 87 and 88.

File: Code/C++/tim_sort.cpp [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    { 
+    // From the start of the array, we are going to merge two segments each of size SIZE.
+        for (int left = 0; left < n; 
+                            left += 2*size) 
+        { 
+            
+            // Find thier mid point
+            int mid = left + size - 1; 
+            int right = min((left + 2*size - 1), 
+                                            (n-1));
```
Comment (raksha009) [inline Code/C++/tim_sort.cpp:94]: plz combine line 93 and 94.

File: Code/C++/tim_sort.cpp [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    // Print the sorted array
+    for (int i = 0; i < n; i++)
+        printf(" %d ", arr[i]);
+    cout << "\n";
+
+    return 0;
+}
+
+/*
+Time Complexity- O(n*log(n))
```
Comment (raksha009) [inline Code/C++/tim_sort.cpp:131]: Add time complexity and space complexity at end.

File: Code/C++/tim_sort.cpp [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+Space Complexity- O(n)
+
+SAMPLE INPUT AND OUTPUT
```
Comment (raksha009) [inline Code/C++/tim_sort.cpp:134]: Donot write word Sample.

**Review by raksha009 — CHANGES_REQUESTED** [review]

**Comment by Harikrishnan6336** [comment]
@raksha009  I have made the suggested changes, Kindly review my PR.

File: Code/C++/tim_sort.cpp [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    for (int i = 0; i < n; i+=SIZE) 
+        insertion_sort(arr, i, min((i+15), (n-1))); 
+    
+    // Merge all the small segments formed
+    for (int size = SIZE; size < n; size = 2*size) 
+    { 
+    // From the start of the array, we are going to merge two segments each of size SIZE.
+        for (int left = 0; left < n; left += 2*size) 
+        { 
+
```
Comment (raksha009) [inline Code/C++/tim_sort.cpp:85]: Remove extra line

File: Code/C++/tim_sort.cpp [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        insertion_sort(arr, i, min((i+15), (n-1))); 
+    
+    // Merge all the small segments formed
+    for (int size = SIZE; size < n; size = 2*size) 
+    { 
+    // From the start of the array, we are going to merge two segments each of size SIZE.
+        for (int left = 0; left < n; left += 2*size) 
+        { 
+            
+            // Find thier mid point
```
Comment (raksha009) [inline Code/C++/tim_sort.cpp:86]: sp. of their -- simply write "find mid point"

File: Code/C++/tim_sort.cpp [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    for (int i = 0; i < n; i++)
+        printf(" %d ", arr[i]);
+    cout << "\n";
+
+    return 0;
+}
+
+/*
+Test Case :
+1.
```
Comment (raksha009) [inline Code/C++/tim_sort.cpp:126]: Write in format of input/ output

File: Code/C++/tim_sort.cpp [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+How many numbers do you want to sort? 5
+Enter the numbers: 1 3 5 2 4
+The numbers in sorted order is:  1  2  3  4  5
+2.
+How many numbers do you want to sort? 0
+There are no numbers to sort!!!
+
+Time Complexity- O(n*log(n))
```
Comment (raksha009) [inline Code/C++/tim_sort.cpp:134]: Specify n

**Review by raksha009 — CHANGES_REQUESTED** [review]

**Comment by raksha009** [comment]
> @anjul2000 @raksha009 I have made the suggested changes, Kindly review my PR. Also My issue was labelled with `gssoc21` and `Level2` labels could you also add these labels to my PR

Don't worry, we'll add the labels and tags before merging yr pr :)

**Comment by Harikrishnan6336** [comment]
I have made the required changes, Kindly review my PR @raksha009

File: Code/C++/tim_sort.cpp [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+
+// Perform insertion sort
+void insertion_sort(int arr[], int beg, int end) 
+{ 
+    for (int i = beg + 1; i <= end; i++) 
+    { 
+        int temp = arr[i]; 
+        int j = i - 1; 
+        
+        /* Shift elements of arr[0..i-1], that are
```
Comment (raksha009) [inline Code/C++/tim_sort.cpp:18]: Reduce it to 2 line comment.

**Review by raksha009 — CHANGES_REQUESTED** [review]

File: Code/C++/tim_sort.cpp [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+It is used in Java’s Arrays.sort() as well as Python’s sorted() and sort()
+*/ 
+#include<bits/stdc++.h> 
+using namespace std; 
+const int SIZE = 16; 
+
+// Perform insertion sort
+void insertion_sort(int arr[], int beg, int end) 
+{ 
+    for (int i = beg + 1; i <= end; i++)
```
Comment (raksha009) [inline Code/C++/tim_sort.cpp:13]: use appropriate variable name, instead of beg, use start.

**Review by raksha009 — CHANGES_REQUESTED** [review]

**Review by raksha009 — APPROVED** [review]
LGTM!👍

**Comment by Harikrishnan6336** [comment]
@tarun26091999 Kindly review my PR.

**Review by rudrakshi99 — APPROVED** [review]

**Pull request merged**